terriajs 2588 server implementation

a-stacey opened the pull request:
This implementation is insecure in that it effectively takes a service which is secured by username and password and allows anyone (everyone) to anonymously mint tokens for this service. This implementation only obfuscates token generation and worse it masks the fact that the layer was secured and now it is not and becomes effectively publicly available (effectively anyone could mint tokens using this service and then connect directly to the server by-passing authentication for as long token is valid). This implementation does not currently explicitly log information about the clients requesting tokens. If this is problematic in the future further steps should be taken to secure this operation such as authenticating the user prior to minting tokens and taking security more seriously such as logging client information and securing connections between the client and esri token server (see more details later).

Please pay particular attention to security considerations and vulnerabilities as I have not written code like this before and so may have made rookie mistakes and overlooked security issues. One thing that does stand out for me is whether I have done enough to sanitize `targetUrl` and then used it like so `tokenServers[targetUrl];` and whether anything malicious could be done here.

This code does not take security seriously as it permits code between both this server and the token server (terriajs-server -> esri token server) and between terriajs and this server (terriajs -> terriajs-server) to be unencrypted (non TLS). Between terriajs-server and the esri token server this allows both the username / password and token in plain text (unencrypted) and so all the information could be used by an attacker. Similarly between terriajs and terriajs-server the token is in plain text and could be intercepted by an attacker (this connection is particularly vulnerable as it is between the client and server and so has increased risk that the connection has opportunity for attack -> i.e. free unencrypted wifi or some such). Ideally the server would enforce that both of these connections be conducted over the most recent version of TLS, however unfortunately the services that we must authenticate do not correctly support TLS and so we are unable to test or enforce this correctly.

kring commented:
😆 let me address some of this.

> This implementation is insecure in that it effectively takes a service which is secured by username and password and allows anyone (everyone) to anonymously mint tokens for this service.

This is partially by design, and partially not true.  It allows anyone that can access terriajs-server to mint tokens, yes.  It's possible for an application to stick authentication in front of terriajs-server though, in which case the token minting is authenticated, too.

But for our use case, no authentication will be required to mint tokens.  We _want_ anonymous users to be able to view the authenticated layers, so there's no way around this.  We're just trying to make it unlikely to accidentally re-use the layers in a different application (because this is not allowed by the license terms), and slightly more difficult to do it purposely (there's no possible way to make it impossible to do purposely).

Calling it "insecure" is unfair, because that implies it's somehow going to leak personal information, allow the server to be compromised, etc, which is definitely not the case here.  This mechanism is closer to DRM; it's helping to enforce someone's copyright.  All DRM is imperfect, but the implementation here is a simple mechanism for discouraging casual misuse of the coprighted material.

> This code does not take security seriously as it permits code between both this server and the token server (terriajs-server -> esri token server) and between terriajs and this server (terriajs -> terriajs-server) to be unencrypted (non TLS).

Again unfair.  It's true that it doesn't require TLS, but that's extremely unlikely to matter.  Plaintext tokens can be sniffed off coffee shop wifi easily enough, but what does it matter when anyone can mint their own token anyway?  If terriajs-server sends the username/password as plaintext to the remote server, that's a bit more concerning.  Ideally the remote server would support HTTPS.  If it does, terriajs-server will use it and that's great.  But even if it doesn't, the potential eavesdroppers between a production terriajs-server and a production ArcGIS server are limited to infrastructure providers or perhaps someone that has compromised an infrastructure provider.  Even in that scenario, there is no leak of personal information or compromise of a server.  It just means the simple DRM scheme is broken.  Layer providers can make their DRM a bit stronger, if they're concerned, by supporting HTTPS.

kring commented on lib/controllers/esri-token-auth.js:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        let targetUrl = parseURL(parameters.url);
+        if (!targetUrl || targetUrl.length === 0 && (typeof targetUrl === string)) {
```
This logic is not right.  `typeof targetUrl` returns `"string"`.  `string` (without quotes) is an undefined symbol and I believe this code will crash if `targetUrl.length` is 0.

Also I believe `&&` has higher precedence than `||`, so this is equivalent to:
```
if (!targetUrl || (targetUrl.length === 0 && typeof targetUrl === string)) {
```
which I don't think is what you intended.

a-stacey replied:
This is terrible code on my behalf... I don't know what I was doing, thanks for picking this up.

e874f7e2064db6fa592d0cc421a6d805a8a273da

kring commented:
> One thing that does stand out for me is whether I have done enough to sanitize targetUrl and then used it like so tokenServers[targetUrl]; and whether anything malicious could be done here.

This is fine, nothing malicious can be done here because `targetUrl` is guaranteed to be a JSON type.

kring commented on lib/controllers/esri-token-auth.js:
```diff
@@ -0,0 +1,86 @@
+        }, function(error, response, body) {
+            res.set('Content-Type', 'application/json');
+
+            if (response.statusCode != 200) {
```
`!==`

a-stacey replied:
Thanks.

bee6bb55f0f4feb506c73ca2c8af23c9585acd1f

kring commented on lib/controllers/esri-token-auth.js:
```diff
@@ -0,0 +1,86 @@
+        });
+    });
+
+    router.use(function(err, req, res, next) {
```
What's the purpose of this?

a-stacey replied:
Not sure whether I understand this question correctly but it is to handle JSON parser errors as well as any other possible errors with the router. Since the JSON is coming from the client it can be completely invalid:
https://expressjs.com/en/guide/error-handling.html

Have added a comment to try and clarify this. 

Can generate an appropriate error using:
`curl -X POST --data "[object]" localhost:3001/esri-token-auth --header "Content-Type:application/json"`

4e24c4757fef983862645350de6773caf2cc7e76

kring replied:
Ok, I don't think it makes sense to handle all errors by returning a generic 400 and "Unknown error" though.  Specific errors (e.g. user provided bad data) should be handled by returning an appropriate 4xx code directly in the main handler.  "Unknown" errors can just be handled by the default express behavior of returning a 500 (internal server error), because an unknown error is the server's fault not the user's.

a-stacey replied:
I'm not exactly sure what you mean here, since I think that this is the only way to get a handle on these errors (they don't seem to propagate through `router.post` since they are generated from the `router.use(bodyParser.json(`).

Take the point on status codes, but still not sure on whether there is a clean way of differentiating them.

kring replied:
doesn't `bodyParser.json()` automatically return a 4xx code then?

a-stacey replied:
Its a `<title>500 - Internal server error</title>`, but I guess this is cleaner.

Have removed.

913312d209ce5d1d5f72d2b6f66ab2baffa02472

kring replied:
500?  that's really frustrating.  If the body is supposed to be JSON and it's not, that's pretty clearly an error on the part of the caller, not an internal server error.  I'd call that a bug in bodyParser.  Anyway, I'm not that concerned about it.

a-stacey replied:
>  I'd call that a bug in bodyParser.

Yeah agreed.

kring commented on lib/controllers/esri-token-auth.js:
```diff
@@ -0,0 +1,86 @@
+    let result = {};
+
+    for(var server in servers) {
+        // Note: We should really validate here that the URL is HTTPS to save us from ourselves,
```
Printing a warning when using a non-https server would be reasonable.

kring commented on lib/controllers/esri-token-auth.js:
```diff
@@ -0,0 +1,86 @@
+function parseUrls(servers) {
+    let result = {};
+
+    for(var server in servers) {
```
You need to either use `if (servers.hasOwnProperty(server)) { ... }` or rewrite it to use `Object.keys(servers)` instead.

a-stacey replied:
Thanks.

97267c4472e6f7bfbd00519a85cadc8f4d8acf2e

kring commented:
Looks good other than those few comments!

a-stacey commented:
Think I have addressed all of the comments in this PR now, with the exception of the 
> Printing a warning when using a non-https server would be reasonable.
which would be nice, but leaving this for now as otherwise running out of time. Might come back to this later and followup with a later PR to patch that and also possibly prevalidate the server config when the config is loaded rather then having it fail when a user trys to access the server.

Please note that I have otherwise changed the checks in the main body quite a bit to clean up some errors and hopefully simplify the logic (double testing things) due to my previous limited understanding.
